### PR TITLE
Bishop

### PR DESCRIPTION
**Pull request by mjhoy**
I ended up writing my own iterator to deal with stepping through ranks/files backwards or forwards. It's a little bit weird because I'm just dealing with `u8` values for the board, so I don't actually ever want to go negative, there are some assertions around this.

### CODE REVIEW

File: src/from_to_step.rs [hunk 1]
```diff
@@ -0,0 +1,60 @@
+pub struct FromToStep {
+    from: u8,
```
Comment (paulkernfeld) [inline src/from_to_step.rs]: Would it be easier to just use `i8` for all of these?

File: src/from_to_step.rs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    to: u8,
+    step: i8,
+}
+
+impl Iterator for FromToStep {
```
Comment (paulkernfeld) [inline src/from_to_step.rs:7]: If you're interested, there are a bunch of cool optional `Iterator` functions that you could implement (for performance), like `size_hint`.

File: src/from_to_step.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    type Item = u8;
+
+    fn next(&mut self) -> Option<u8> {
+        let new_next: i8 = (self.from as i8) + self.step;
+
+        if self.step < 0 {
+            assert!(self.from > 0, "from cannot go negative");
```
Comment (paulkernfeld) [inline src/from_to_step.rs]: Maybe this should be a `debug_assert`, since it seems pretty impossible.

File: src/from_to_step.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        if self.from >= self.to && self.step > 0 {
+            return None;
+        }
+        Some(self.from)
+    }
+}
+
+impl FromToStep {
+    pub fn from_to(from: u8, to: u8) -> FromToStep {
+        assert!(from != to, "from and to must be different values");
```
Comment (paulkernfeld) [inline src/from_to_step.rs]: There is an `assert_ne` macro that could be used for this.

File: src/from_to_step.rs (hunk 1 above)
Comment (mjhoy) [reply src/from_to_step.rs]: yes, good point. 56aa5b9cb66d36dfed6d3a33cc1e78dd06c5234c

File: src/from_to_step.rs (hunk 2 above)
Comment (mjhoy) [reply src/from_to_step.rs:7]: cool. i think i'm doing this one right? 4404ceabaaf9fd2597bea2e264f5614837f80db7

File: src/from_to_step.rs (hunk 3 above)
Comment (mjhoy) [reply src/from_to_step.rs]: ah. i actually hadn't thought about the fact that the fields in `FromToStep` couldn't get set directly, so yeah that's true! 46d85374c2e366489b705499e9170b32f911b97e

File: src/from_to_step.rs (hunk 4 above)
Comment (mjhoy) [reply src/from_to_step.rs]: nice! 8eb97f1f4f8cc4151cd3917811804769ad98e6fa

**Comment by mjhoy** [comment]
thanks @paulkernfeld !

**Pull request merged**